x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DataTierMigrationRoutedStep.java

```diff
@@ -0,0 +1,119 @@
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_EXCLUDE_SETTING);
+        ALL_CLUSTER_SETTINGS = allSettings;
+    }
+
+    private static final AllocationDeciders ALLOCATION_DECIDERS = new AllocationDeciders(
+        List.of(
+            new DataTierAllocationDecider(new ClusterSettings(Settings.EMPTY, ALL_CLUSTER_SETTINGS))
+        )
+    );
+
+    DataTierMigrationRoutedStep(StepKey key, StepKey nextStepKey) {
+        super(key, nextStepKey);
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        IndexMetadata idxMeta = clusterState.metadata().index(index);
+        if (idxMeta == null) {
+            // Index must have been since deleted, ignore it
+            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), index.getName());
+            return new Result(false, null);
+        }
+        String destinationTier = INDEX_ROUTING_INCLUDE_SETTING.get(idxMeta.getSettings());
+        if (ActiveShardCount.ALL.enoughShardsActive(clusterState, index.getName()) == false) {
+            logger.debug("[{}] migration of index [{}] to the [{}] tier cannot progress, as not all shards are active",
+                    getKey().getAction(), index.getName(), destinationTier);
+            return new Result(false, AllocationInfo.waitingForActiveShardsAllocationInfo(idxMeta.getNumberOfReplicas()));
+        }
+
+        int allocationPendingAllShards = getPendingAllocations(index, ALLOCATION_DECIDERS, clusterState);
+
+        if (allocationPendingAllShards > 0) {
+            boolean targetTierNodeFound = false;
+            for (DiscoveryNode node : clusterState.nodes()) {
+                for (DiscoveryNodeRole role : node.getRoles()) {
+                    if (role.roleName().equals(DATA_ROLE.roleName()) || role.roleName().equals(destinationTier)) {
+                        targetTierNodeFound = true;
+                        break;
+                    }
+                }
+            }
+            String statusMessage = String.format(Locale.ROOT, "%s lifecycle action [%s] waiting for [%s] shards to be moved to the [%s] " +
+                    "tier" + (targetTierNodeFound ? "" : " but there are currently no [%s] nodes in the cluster"),
+                index, getKey().getAction(), allocationPendingAllShards, destinationTier, destinationTier);
+            logger.debug(statusMessage);
+            return new Result(false, new AllocationInfo(idxMeta.getNumberOfReplicas(), allocationPendingAllShards, true, statusMessage));
+        } else {
+            logger.debug("[{}] migration of index [{}] to tier [{}] complete", getKey().getAction(), index, destinationTier);
+            return new Result(true, null);
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return 711;
```

Comment:
I think this needs to incorporate `super.hashCode()` or else the hash code will be the same regardless of stepkeys